File: Solutions/Python/snakes_and_ladders.py
"""
    [MEDIUM]
    909. Snakes and Ladders

    Concepts:
    - matrix
    - breadth-first search

    Stats:
        Runtime | 24 ms     [Beats 47.52%]
        Memory  | 17.96 MB  [Beats 59.05%]
"""

class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)
        min_rolls = float("inf")
        n_squared = n ** 2

        def to_coords(idx: int) -> tuple[int]:
            return (
                (n - 1 - (idx // n)),
                (n - 1 - (idx % n))
                if (idx // n) & 1
                else (idx % n)
            )

        # convert indices to boustrophedon coordinates
        idx_to_coords, coords_to_idx = {}, {}
        for i in range(n_squared):
            idx_to_coords[i+1] = to_coords(i)
            coords_to_idx[to_coords(i)] = i+1
        
        # explore board
        def bfs():
            queue = deque([(1, 0)])  # idx, moves
            visited = set([1])

            while queue:
                len_q = len(queue)

                for _ in range(len_q):
                    idx, moves = queue.popleft()

                    if idx == n_squared:
                        return moves

                    for didx in reversed(range(1, 7)):
                        nidx = idx + didx
                        if nidx > n_squared:
                            continue

                        i, j = idx_to_coords[nidx]
                        if board[i][j] != -1:
                            nidx = board[i][j]

                        if nidx not in visited:
                            queue.append((nidx, moves + 1))
                            visited.add(nidx)

            return -1

        return bfs()

File: Solutions/Cpp/snakes_and_ladders.cpp
/*
    [MEDIUM]
    909. Snakes and Ladders

    Concepts:
    - matrix
    - breadth-first search

    Stats:
        Runtime | 5 ms      [Beats 42.00%]
        Memory  | 22.41 MB  [Beats 9.29%]
*/

class Solution {
public:
    int n;

    std::array<int, 2> to_coords(int idx) {
        return (
            std::array<int, 2>{
                (n - 1 - (idx / n)),
                ((idx / n) & 1) ?
                (n - 1 - (idx % n)) :
                (idx % n)
            }
        );
    }

    int snakesAndLadders(vector<vector<int>>& board) {
        n = board.size();
        int n_squared = std::pow(n, 2);

        // convert indices to boustrophedon coordinates
        std::unordered_map<int, std::array<int, 2>> idx_to_coords = {};
        for (int i = 0; i < n_squared; i++)
            idx_to_coords[i+1] = to_coords(i);

        // explore board
        std::function<int(void)> bfs;

        bfs = [&]() {
            std::deque<std::pair<int, int>> dq;
            dq.push_back({1, 0});
            std::unordered_set<int> visited = {1};
            int len_q, idx, nidx, moves, i, j;

            while (!dq.empty()) {
                len_q = dq.size();

                for (int _ = 0; _ < len_q; _++) {
                    auto [idx, moves] = dq.front();
                    dq.pop_front();

                    if (idx == n_squared)
                        return moves;

                    for (int didx = 6; didx > 0; didx--) {
                        nidx = idx + didx;
                        if (nidx > n_squared)
                            continue;

                        auto [i, j] = idx_to_coords[nidx];
                        if (board[i][j] != -1)
                            nidx = board[i][j];

                        if (!visited.contains(nidx)) {
                            dq.push_back({nidx, moves + 1});
                            visited.insert(nidx);
                        }
                    }
                }
            }

            return -1;
        };

        return bfs();
    }
};